src/main/java/org/datadog/jenkins/plugins/datadog/clients/CompressedBatchSender.java

```diff
@@ -0,0 +1,69 @@
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.logging.Logger;
+import java.util.zip.GZIPOutputStream;
+
+public class CompressedBatchSender<T> implements JsonPayloadSender<T> {
+
+    private static final Logger logger = Logger.getLogger(CompressedBatchSender.class.getName());
+
+    private static final byte[] BEGIN_JSON_ARRAY = "[".getBytes(StandardCharsets.UTF_8);
+    private static final byte[] END_JSON_ARRAY = "]".getBytes(StandardCharsets.UTF_8);
+    private static final byte[] COMMA = ",".getBytes(StandardCharsets.UTF_8);
+
+    private final HttpClient httpClient;
+    private final String url;
+    private final Map<String, String> headers;
+    private final int batchLimitBytes;
+    private final Function<T, String> payloadToString;
+
+    public CompressedBatchSender(HttpClient httpClient,
+                                 String url,
+                                 Map<String, String> headers,
+                                 int batchLimitBytes,
+                                 Function<T, String> payloadToString) {
+        this.httpClient = httpClient;
+        this.url = url;
+        this.headers = headers;
+        this.batchLimitBytes = batchLimitBytes;
+        this.payloadToString = payloadToString;
+    }
+
+    @Override
+    public void send(Collection<T> payloads) throws Exception {
+        ByteArrayOutputStream request = new ByteArrayOutputStream();
+        GZIPOutputStream gzip = new GZIPOutputStream(request);
+        // the backend checks the size limit against the uncompressed body of the request
+        int uncompressedRequestLength = 0;
+
+        for (T payload : payloads) {
+            byte[] body = payloadToString.apply(payload).getBytes(StandardCharsets.UTF_8);
+            if (body.length + 2 > batchLimitBytes) { // + 2 is for array beginning and end: [<payload>]
+                logger.severe("Dropping a payload because size (" + body.length + ") exceeds the allowed limit of " + batchLimitBytes);
+                continue;
+            }
+
+            if (uncompressedRequestLength + body.length + 2 > batchLimitBytes) { // + 2 is for comma and array end: ,<payload>]
+                gzip.write(END_JSON_ARRAY);
+                gzip.close();
+                httpClient.post(url, headers, "application/json", request.toByteArray(), Function.identity());
+                request = new ByteArrayOutputStream();
+                gzip = new GZIPOutputStream(request);
+                uncompressedRequestLength = 0;
+            }
+
+            gzip.write(uncompressedRequestLength == 0 ? BEGIN_JSON_ARRAY : COMMA);
+            gzip.write(body);
+            uncompressedRequestLength += body.length + 1;
+        }
+
+        gzip.write(END_JSON_ARRAY);
+        gzip.close();
+        httpClient.post(url, headers, "application/json", request.toByteArray(), Function.identity());
```

Comment:
I could not use synchronous post in the past as there was no batching. Sending traces one by one would be very slow, so I just relied on async requests to be done in parallel. Now that we have batching they can be sent sequentially, rather than in parallel. The advantage is that we don't open extra connections to the same host (TCP works better when there are fewer connections per endpoint, otherwise many connections are in each other's way when doing congestion control).